### ISSUE TITLE
Feature request: failure summary

### ISSUE THREAD

**Issue by gmile** [rspec/rspec-core#1913]
In a project of mine I'm seeing a flickering test, which is very hard to reproduce. I would be super handy to have access to some expended test metadata, like:

1. list of objects, exposed by `let` and `subject` helpers,
2. list of registered mocks and messages on them, allowed to be called.

**It would be just great to store & have access to failure metadata nested under `example`s**. Also not necessary, but would be cool to have an option to dump this alongside the failure output, like this:

```
3) Api::V1::ProductsController Emplyoees should be able to edit products
     Failure/Error: expect(product.reload.name).to eq('test name')
       
       expected: "updated name"
            got: "original name"
       
       (compared using ==)

     Failure summary:

       Subject:
         Api::V1::ProductsController

       Helpers:
         product: #<Product id: 1, vendor_id: 2 ...>
         vendor: #<Vendor id: 2 ...>

       Mocks:
         employee: double('Employee')
           allowed messages: 
             a: nil
             b: return_result_y
             c: return_result_z
           called messages: :a, :c
         some_object: double('SomeOject')
           allowed messages: 
             p: return_result_l
             q: nil
             v: return_result_p
           called messages:
```

**Comment by myronmarston** [comment]
Interesting idea, @gmile!  I think this is pretty doable.  Are you interested in submitting a PR to add APIs to access an examples `let` declarations and test doubles?

**Comment by gmile** [comment]
Hi @myronmarston, yes absolutely. Will start looking into this weekend. I'm barely familiar with rspec code base. Any guidance, even if high-level, would be highly appreciated!

**Comment by myronmarston** [comment]
Here's how you can get all let/subject declarations:

``` ruby
module RSpec
  module Core
    class Example
      def let_declarations
        @let_declarations ||= FlatMap.flat_map(@example_group_class.parent_groups) do |g|
          MemoizedHelpers.module_for(g).instance_methods(false)
        end
      end
    end
  end
end
```

Which is available to be called like so:

``` ruby
RSpec.describe SomeClass do
  let(:foo) { ... }
  let(:bar) { ... }

  it 'can access the let declarations' do |ex|
    ex.let_declarations # => [:foo, :bar]
  end
end
```

The example object is also passed to the formatter so a formatter could use it as well.

For rspec-mocks test doubles, we'll need to add an API to rspec-mocks for it.  Bear in mind that rspec-core, rspec-expectations and rspec-mocks are all designed to be usable individually in isolation.

**Comment by danascheider** [comment]
@gmile - are you still working on this? Is there any way I can help? This would be really useful to me as well.

**Comment by gmile** [comment]
Hi @danascheider, no I was not able to start work on this, unfortunately. You can take this up.

**Comment by danascheider** [comment]
Will do.

**Comment by danascheider** [comment]
@myronmarston, how do you feel about adding a `--debug` switch that would turn this functionality on? I can imagine a situation where calling it that would be problematic, so if you'd like me to call it something else - or turn the functionality on through some other means, like config settings - just let me know what you think.

**Comment by JonRowe** [comment]
Personally, given that this is essentially just extra output, I would suggest that this is implemented as a formatter, then it can be triggered from the command line, `.rspec` or config. This would also allow it to be implemented as a third party gem for now (although we'll happily help by adding APIs to access this information as Myron has said)

**Comment by benoittgt** [comment]
Love the idea to use this as a formatter.

**Comment by danascheider** [comment]
Sounds good, I'll go ahead and do it that way.

**Comment by myronmarston** [comment]
I agree that a formatter makes sense.  I'm also not sure that such a formatter belongs in rspec-core but any APIs that formatter would need (such as `let_declarations`) certainly would.
